test/test_integration_pr2.py

```diff
@@ -331,6 +332,40 @@ def test_prismatic_joint1(self, zero_pose):
         js = {u'torso_lift_joint': 0.1}
         zero_pose.send_and_check_joint_goal(js)
 
+    def test_hard_joint_limits(self, zero_pose):
+        """
+        :type zero_pose: PR2
+        """
+        zero_pose.allow_self_collision()
+        goal_js = {u'r_elbow_flex_joint': -2.4,
+                   u'torso_lift_joint': 0.4,
+                   u'head_pan_joint': 3.5}
+        zero_pose.set_joint_goal(goal_js)
+        zero_pose.send_goal()
+        js = rospy.wait_for_message('/joint_states', JointState)
+        torso_lift_joint_state = js.position[js.name.index('torso_lift_joint')]
+        r_elbow_flex_joint_state = js.position[js.name.index('r_elbow_flex_joint')]
+        head_pan_joint_state = js.position[js.name.index('head_pan_joint')]
+
+        assert(torso_lift_joint_state <= 0.32500000000000007 and torso_lift_joint_state >= 0.0115)
```

Comment:
- Please use `zero_pose.get_robot().get_joint_limits`, which returns a tuple with lower and upper limit. That way you can also easily set goals and check the upper and lower limits of all non continuous joints.
- There is `zero_pose.get_trajectory_msg`, which returns a list of dicts that maps joint name to position. Please rename that function to `get_result_trajectory_position`, create a similar one for velocity and then check the whole trajectory for constraint violations. This should probably also be a function in the class where `get_trajectory_msg` is. Call it `are_joint_limits_violated` or something.
- To deal with the precision error, round the numbers to 7 decimal places before the check.